### PR TITLE
Final PR

### PR DESCRIPTION
**Pull request by quinlanbock97**
Deleted src files in one branch and created PR from final branch to this branch to see all src final changes.

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+const app = express();
+const port = 3000;
+
+
+app.use(express.json()) // for parsing application/json
+app.use(express.urlencoded({ extended: true })) // for parsing application/x-www-form-urlencoded
+
+app.get('/', function (req, res) {
+    const listDes = [];
+    if (MyTodo.list.length !== 0) {
```
Comment (stieg) [inline index.ts]: So this would be really hard to parse from an API point of view.  It would be simply be better to return an empty array to indicate that there are no items no?

Comment (stieg) [reply index.ts]: With that in mind I think you will find that you can simplify this method

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+            listDes.push(MyTodo.list[i].description)
+        }
+        res.json(listDes);
+    } else {
+        res.json('No items in list.');
+    }
+});
+
+app.post('/', function (req, res) {
+    const todoItem = MyTodo.addItem(req.body.Description, req.body.Date, req.body.IsDone);
```
Comment (stieg) [inline index.ts]: So this is dubious.  What happens if the caller doesn't provide a Description, date or isDone value?  What if it is malcious?  You must always sanitize and validate user input before using it.  Can not stress that enough in any situation where you take info from an external (and untrusted) source

Comment (stieg) [reply index.ts]: This applies throughout to everything that a user can specify.  Think URL params, HTTP body content, all of it.

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+        res.json(todoItem);
+    } else {
+        res.sendStatus(400);
+    }
+});
+
+app.put('/:id/done', function (req, res) {
+    var success = MyTodo.markDone(req.params.id);
+    if (success) {
+        res.sendStatus(200);
```
Comment (stieg) [inline index.ts]: Fun way to make this shorter: `res.sendStatus(success ? 200 : 400);`

Comment (stieg) [reply index.ts]: Ternary statements are awesome.  Check them out if you are not familiar with them.

File: src/todo.test.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
+﻿import { Todo } from "./todo";
+import { ListItem } from "./todo";
+const MyTodo = new Todo();
+
+beforeAll(() => {
```
Comment (stieg) [inline src/todo.test.ts:5]: Good use of a setup method here.

File: src/todo.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+* Date to due by
+* Unique ID for the task
+* Whether the task is done or not
+*/
+export interface ListItem {
+    createdAt: Date;
+    description: string;
+    dueAt: Date;
+    id: number;
+    isDone: boolean;
```
Comment (stieg) [inline src/todo.ts]: So more of a personal preference here but I often use Date objects in place of booleans since Dates not only act like booleans (being set vs unset) but they also tell you when.  This is especially useful when it comes to making objects that get stored in a DB.

File: src/todo.ts [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+    public list: ListItem[];
+    public idNumber: number;
+    constructor() {
+        this.list = [];
+        this.idNumber = 0;
+    }
+
+    // Function to add items
+    public addItem(des: string, dateString: string, done: boolean = false) {
+        const today = new Date();
```
Comment (stieg) [inline src/todo.ts]: Not really needed as a separate variable here since it isn't used more than once.  Can simply move it down to line 32.

File: src/todo.ts [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        const dueDate = new Date(dateString);
+        if (dueDate instanceof Date && !isNaN(dueDate.valueOf())) {
```
Comment (stieg) [inline src/todo.ts]: Consider saving indentation by checking for a failure scenario only.  So like this:
```
if (is not a Date) {
  throw new Error('not a date');
}
...
```
See how this saves a lot of indentation.  And if the check fails the throw stop execution of the function.

File: src/todo.ts [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+    createdAt: Date;
+    description: string;
+    dueAt: Date;
+    id: number;
+    isDone: boolean;
+}
+
+// class that is use to create todo lists
+export class Todo {
+    public list: ListItem[];
```
Comment (stieg) [inline src/todo.ts]: So do you really want this list to be public?  If so, why?  Check all internal state variables here against this.

File: src/todo.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+        } else {
+            throw new Error("The date entered is not valid.");
+        }
+    }
+
+    // Function to remove an Item given the ID
+    public removeItem(id: number): ListItem {
+        var item: ListItem;
+        const len: number = this.list.length;
+        for (let i = 0; i < len; i++) {
```
Comment (stieg) [inline src/todo.ts]: So this works.  But there are some very nice array methods that make this much simpler.  Lookup `Array.find`.

File: src/todo.ts [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+        }
+        return item;
+    }
+
+    // Prints out Info on each TODO item
+    public getListItems(): number[] {
+        const listIDs: number[] = [];
+        for (const item of this.list) {
+            const stat: string = (item ? "Completed" : " Not completed");
+            listIDs.push(item.id);
```
Comment (stieg) [inline src/todo.ts]: Look up the function `Array.map`.  May simplify things much more.

File: src/todo.ts [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+            console.log(
+                "ID: " + item.id +
+                ", Description: " + item.description +
+                ", Created Date: " + (item.createdAt.getMonth() + 1) + "/" +
```
Comment (stieg) [inline src/todo.ts]: Notice that you are repeating code here.  Maybe it would be easier to create a method that returns that string when given a date?  Always keep an eye out for things like this. Remember the DRY principle (Don't Repeat Yourself).

File: src/todo.ts [hunk 12]
```diff
@@ -0,0 +1,123 @@
...
+                                  item.dueAt.getHours() + ":00" +
+                "Status: " + stat);
+        }
+        return listIDs;
+    }
+
+    // Prints out Info on each TODO item
+    public getCompletedListItems(): number[] {
+        const listIDs: number[] = [];
+        for (const item of this.list) {
```
Comment (stieg) [inline src/todo.ts]: Look up `Array.filter`.  Combined with `Array.map` will make this super simple.  Just a thought.

File: src/todo.ts [hunk 13]
```diff
@@ -0,0 +1,123 @@
...
+            if (item.isDone) {
+                listIDs.push(item.id);
+                console.log("ID: " + item.id + ", Description: " + item.description);
+            }
+        }
+        return listIDs;
+    }
+
+    // Prints out Info on each TODO item
```
Comment (stieg) [inline src/todo.ts]: Going forward you should always be using tsdoc style docs for these descriptions.  This way IDEs and documentation can be generated.

File: src/todo.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+    // Clear all TODOs
+    public clear(): void {
+        this.list = [];
+    }
+
+    // Mark as done
+    public markDone(id: number): boolean{
+        var success = false;
+        const len: number = this.list.length;
+        for (let i = 0; i < len; i++) {
```
Comment (stieg) [inline src/todo.ts]: You have repeated your find the item function twice.  This is not ideal.  You should use a private function to do this find and call it from the places you need it. Also is there a faster way to lookup an object instead of using a full array scan?  Think about how your data is being stored in the array type.

File: src/todo.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+            if (this.list[i].id == id) {
+                this.list[i].isDone = true;
+                success = true;
+                break;
+            }
+        }
+        return success;
```
Comment (stieg) [inline src/todo.ts]: So if its not found should it be false or an Error?  I personally would throw an error since it would imply the object with the given id was not found.

**Review by stieg — COMMENTED** [review]
So be very careful when it comes to trusting user input with web apps.  Most people are good but there are jerks out there that will hack an app that does not sanitize input.

File: index.ts (hunk 2 above)
Comment (quinlanbock97) [reply index.ts]: So I already had functionality to check if either weren't entered and what I believe correctly handled that. I was going to compartmentalize this and checks for malicious things in a function. What sort of malicious things might I check for. I am aware of nasty stuff like this in SQL, but not so much in JS. Would i look for things like execute, \n, and such being imputed?

File: src/todo.ts (hunk 8 above)
Comment (quinlanbock97) [reply src/todo.ts]: I wanted list to be public so that instantiated objects would have access to the list. It could also be private with a function that returns it. Is there any difference? Always been confused about this.
